chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -78,640 +112,440 @@ impl ConstrainedValueAccumulator {
+                    self.timelock_constrained.entry(maturity_distance).or_insert(Amount::ZERO);
+                *balance = (*balance + pledged_amount)
+                    .ok_or(IOPolicyError::CoinOrTokenOverflow(CoinOrTokenId::Coin))?;
+            }
+        };
+
+        Ok(())
+    }
+
+    fn process_input_account<DelegationBalanceGetterFn>(
+        &mut self,
+        chain_config: &ChainConfig,
+        block_height: BlockHeight,
+        account: &AccountSpending,
+        delegation_balance_getter: &DelegationBalanceGetterFn,
+    ) -> Result<(), IOPolicyError>
+    where
+        DelegationBalanceGetterFn: Fn(DelegationId) -> Result<Option<Amount>, IOPolicyError>,
+    {
+        match account {
+            AccountSpending::DelegationBalance(delegation_id, spend_amount) => {
+                let delegation_balance = delegation_balance_getter(*delegation_id)?
+                    .ok_or(IOPolicyError::DelegationBalanceNotFound(*delegation_id))?;
+                ensure!(
+                    *spend_amount <= delegation_balance,
+                    IOPolicyError::AttemptToPrintMoney(CoinOrTokenId::Coin)
+                );
+
+                let maturity_distance: BlockCount = chain_config
+                    .staking_pool_spend_maturity_distance(block_height)
+                    .to_int()
+                    .try_into()
+                    .map_err(|_| IOPolicyError::NegativeMaturityDistance)?;
+
+                let balance =
+                    self.timelock_constrained.entry(maturity_distance).or_insert(Amount::ZERO);
+                *balance = (*balance + *spend_amount)
+                    .ok_or(IOPolicyError::CoinOrTokenOverflow(CoinOrTokenId::Coin))?;
+            }
+        };
 
         Ok(())
     }
 
+    fn process_input_account_command(
+        &mut self,
+        chain_config: &ChainConfig,
+        command: &AccountCommand,
+    ) -> Result<Amount, IOPolicyError> {
+        match command {
+            AccountCommand::MintTokens(token_id, amount) => {
+                insert_or_increase(
+                    &mut self.unconstrained_value,
+                    CoinOrTokenId::TokenId(*token_id),
+                    *amount,
+                )?;
+                Ok(chain_config.token_min_supply_change_fee())
+            }
+            AccountCommand::LockTokenSupply(_) | AccountCommand::UnmintTokens(_) => {
+                Ok(chain_config.token_min_supply_change_fee())
+            }
+            AccountCommand::FreezeToken(_, _) | AccountCommand::UnfreezeToken(_) => {
+                Ok(chain_config.token_min_freeze_fee())
+            }
+            AccountCommand::ChangeTokenAuthority(_, _) => {
+                Ok(chain_config.token_min_change_authority_fee())
+            }
+        }
+    }
+
     pub fn process_outputs(
         &mut self,
         chain_config: &ChainConfig,
+        block_height: BlockHeight,
         outputs: &[TxOutput],
     ) -> Result<(), IOPolicyError> {
```

Comment:
I'm wondering if the following approach would help:

The `process_outputs` function returns a value accumulator that represents when are coins/tokens spent and at what is the amount. Then we have inputs accumulator and outputs accumulator and can compare the "spending schedule" of both. I think it would make the checking algorithm simpler.

Just iterate over the amounts in the time lock order and keep a running total of "released" - "outputted" amounts and check that at no point it drops below zero. That works because from the accumulator, it is easy to get the amounts in the order in which the time locks are released.

Reply:
So the last commit `2770982` implements this approach. Looks good, what's your thoughts?